src/utilities/cache-eviction-policy.ts

```diff
@@ -0,0 +1,139 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an 'AS IS' BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * A mutable cache is any object that has that allows cache
+ * items to be deleted imperatively given their key
+ */
+export interface MutableCache<T> {
+  delete(key: T): void;
+}
+
+const $retainerCount = Symbol('retainerCount');
+const $recentlyUsed = Symbol('recentlyUsed');
+const $evict = Symbol('evict');
+const $evictionThreshold = Symbol('evictionThreshold');
+const $cache = Symbol('cache');
+
+/**
+ * The CacheEvictionPolicy manages the lifecycle for items in a cache,
+ * evicting any items outside some threshold bounds in "recently used" order,
+ * if they are evictable.
+ *
+ * Items are considered cached as they are retained. When all retainers
+ * of an item release it, that item is considered evictable.
+ */
+export class CacheEvictionPolicy<T = string> {
+  private[$retainerCount] = new Map<T, number>();
+  private[$recentlyUsed]: Array<T> = [];
+  private[$evictionThreshold]: number;
+  private[$cache]: MutableCache<T>;
+
+  constructor(cache: MutableCache<T>, evictionThreshold: number = 5) {
+    this[$cache] = cache;
+    this[$evictionThreshold] = evictionThreshold;
+  }
+
+  /**
+   * The eviction threshold is the maximum number of items to hold
+   * in cache indefinitely. Items within the threshold (in recently
+   * used order) will continue to be cached even if they have zero
+   * retainers.
+   */
+  set evictionThreshold(value: number) {
+    this[$evictionThreshold] = value;
+    this[$evict]();
+  }
+
+  get evictionThreshold(): number {
+    return this[$evictionThreshold];
+  }
+
+  /**
+   * A reference to the cache that operates under this policy
+   */
+  get cache(): MutableCache<T> {
+    return this[$cache];
+  }
+
+  /**
+   * Given an item key, returns the number of retainers of that item
+   */
+  retainerCount(key: T): number {
+    return this[$retainerCount].get(key) || 0;
+  }
+
+  /**
+   * Resets the internal tracking of cache item retainers. Use only in cases
+   * where it is certain that all retained cache items have been accounted for!
+   */
+  reset() {
+    this[$retainerCount].clear();
+    this[$recentlyUsed] = [];
+  }
+
+  /**
+   * Mark a given cache item as retained, where the item is represented
+   * by its key. An item can have any number of retainers.
+   */
+  retain(key: T) {
+    if (!this[$retainerCount].has(key)) {
+      this[$retainerCount].set(key, 0);
+    }
+    this[$retainerCount].set(key, this[$retainerCount].get(key)! + 1);
+
+    const recentlyUsedIndex = this[$recentlyUsed].indexOf(key);
+
+    if (recentlyUsedIndex !== -1) {
+      this[$recentlyUsed].splice(recentlyUsedIndex, 1);
+    }
+
+    this[$recentlyUsed].unshift(key);
```

Comment:
These two lines seem to be vaguely duplicate? Also, does the $recentlyUsed array grow without bound? Could we just make $recentlyUsed the same length as $retainerCount and just swap elements to the beginning? If only JS had a priority queue...

Reply:
The JavaScript `Array` is unbounded and can grow indefinitely, runtime permitting. One cannot limit the size of an `Array`. JavaScript `Array` is more like a map than an actual array, where the indices are the keys. You can overload its `push` / `pop` etc. methods, but you cannot overload the index operator, and that allows you to insert to arbitrary indices (e.g., `this[$recentlyUsed][49283] = true` will totally do a thing here, and grow the reported `length` accordingly).

`[$retainerCount]` being a `Map`, it is more expensive to count the keys than to just do the book keeping directly as we retain new keys. So, surgically splicing elements out of the `Array` is probably the best we can do.

Reply:
Got it, see above.